src/renderer/html_handlebars/search.rs

```diff
@@ -0,0 +1,221 @@
+    debug!("Copying search files ✓");
+
+    Ok(())
+}
+
+fn make_doc_ref(anchor_base: &str, section_id: Option<String>) -> Cow<str> {
+    if let Some(id) = section_id {
+        format!("{}#{}", anchor_base, id).into()
+    } else {
+        anchor_base.into()
+    }
+}
+
+/// Renders markdown into flat unformatted text and adds it to the search index.
+fn render_item(index: &mut Index, search_config: &Search, item: &BookItem) -> Result<()> {
+    let chapter = match item {
+        &BookItem::Chapter(ref ch) => ch,
+        _ => return Ok(()),
+    };
+
+    let filepath = Path::new(&chapter.path).with_extension("html");
+    let filepath = filepath
+        .to_str()
+        .chain_err(|| "Could not convert HTML path to str")?;
+    let anchor_base = utils::fs::normalize_path(filepath);
+
+    let mut opts = Options::empty();
+    opts.insert(OPTION_ENABLE_TABLES);
+    opts.insert(OPTION_ENABLE_FOOTNOTES);
+    let p = Parser::new_ext(&chapter.content, opts);
+
+    let mut in_header = false;
+    let max_section_depth = search_config.heading_split_level as i32;
+    let mut section_id = None;
+    let mut heading = String::new();
+    let mut body = String::new();
+    let mut breadcrumbs = chapter.parent_names.clone();
+    let mut footnote_numbers = HashMap::new();
+
+    for event in p {
+        match event {
+            Event::Start(Tag::Header(i)) if i <= max_section_depth => {
+                if heading.len() > 0 {
+                    // Section finished, the next header is following now
+                    // Write the data to the index, and clear it for the next section
+                    let doc_ref = make_doc_ref(&anchor_base, section_id);
+
+                    index.add_doc(&doc_ref, &[&heading, &body, &breadcrumbs.join(" » ")]);
+                    section_id = None;
+                    heading.clear();
+                    body.clear();
+                    breadcrumbs.pop();
+                }
+
+                in_header = true;
+            }
+            Event::End(Tag::Header(i)) if i <= max_section_depth => {
+                in_header = false;
+                section_id = Some(utils::id_from_content(&heading));
+                breadcrumbs.push(heading.clone());
+            }
+            Event::Start(Tag::BlockQuote)
+            | Event::End(Tag::BlockQuote)
+            | Event::Start(Tag::CodeBlock(_))
+            | Event::End(Tag::CodeBlock(_))
+            | Event::Start(Tag::Header(_))
+            | Event::End(Tag::Header(_))
+            | Event::End(Tag::Paragraph)
+            | Event::Start(Tag::Rule)
+            | Event::End(Tag::TableHead)
+            | Event::End(Tag::TableCell)
+            | Event::Start(Tag::List(_))
+            | Event::End(Tag::Item)
+            | Event::Start(Tag::Image(..))
+            | Event::End(Tag::Image(..))
+            | Event::SoftBreak
+            | Event::HardBreak => {
```

Comment:
This match statement looks like it was painful and error-prone to write out. Unfortunately I can't think of a better way of expressing this than what you've already got :disappointed:

Reply:
It was trial-and-error, yes, but the good thing is that if it's wrong it doesn't matter that much. Now that I think about it the match could be replaced with `Event::Start(_) | Event::End(_) | Event::SoftBreak | Event::HardBreak` and the only difference would be a few more spurious spaces in the search result teasers. Hmm, and we could collapse multiple spaces down to one.